packages/ui/src/molecules/choose-modal/ft-modal.tsx

```diff
@@ -0,0 +1,177 @@
+import { Principal } from "@dfinity/principal"
+import clsx from "clsx"
+import { useCallback, useEffect, useMemo, useRef, useState } from "react"
+import { IoIosSearch } from "react-icons/io"
+
+import { ChooseTokenSkeleton, IconCmpWarning } from "@nfid-frontend/ui"
+import { Input } from "@nfid-frontend/ui"
+import { IconCmpArrow, Label, Tooltip } from "@nfid-frontend/ui"
+
+import { FT } from "frontend/integration/ft/ft"
+
+import { useIntersectionObserver } from "../../organisms/send-receive/hooks/intersection-observer"
+import { getUserPrincipalId } from "../../organisms/tokens/utils"
+import { ChooseItem } from "./choose-ft-item"
+
+const INITED_TOKENS_LIMIT = 6
+
+export interface IChooseFtModal {
+  options: FT[]
+  preselectedValue?: string
+  onSelect?: (value: string) => void
+  warningText?: string | JSX.Element
+  label?: string
+  title: string
+  trigger?: JSX.Element
+  isSmooth?: boolean
+}
+
+export const ChooseFtModal = ({
+  options,
+  preselectedValue,
+  onSelect,
+  warningText,
+  title,
+  label,
+  trigger,
+  isSmooth = false,
+}: IChooseFtModal) => {
+  const [searchInput, setSearchInput] = useState("")
+  const [isModalVisible, setIsModalVisible] = useState(false)
+  const [selectedValue, setSelectedValue] = useState(preselectedValue ?? "")
+  const [tokensOptions, setTokensOptions] = useState<FT[]>([])
+  const [isTokenOptionsLoading, setIsTokenOptionsLoading] = useState(false)
+  const itemRefs = useRef<(HTMLDivElement | null)[]>([])
+
+  useEffect(() => {
+    const init = async () => {
+      setIsTokenOptionsLoading(true)
+      const { publicKey } = await getUserPrincipalId()
+
+      const tokenOptions = await Promise.all(
+        options.map((token, index) => {
+          return index < INITED_TOKENS_LIMIT
+            ? token.init(Principal.fromText(publicKey))
+            : token
+        }),
+      )
+
+      setTokensOptions(tokenOptions)
+      setIsTokenOptionsLoading(false)
+    }
+
+    init()
+  }, [])
+
+  const filteredTokens = useMemo(() => {
+    if (searchInput.length < 2) return tokensOptions
+
+    return tokensOptions.filter(
+      (token) =>
+        token
+          .getTokenSymbol()
+          .toLowerCase()
+          .includes(searchInput.toLowerCase()) ||
+        token.getTokenName().toLowerCase().includes(searchInput.toLowerCase()),
+    )
+  }, [tokensOptions, searchInput])
+
+  useIntersectionObserver(itemRefs.current, async (index) => {
+    const originalIndex = options.findIndex(
+      (token) => token === filteredTokens[index],
+    )
+
+    if (originalIndex !== -1 && !options[originalIndex].isInited()) {
+      const { publicKey } = await getUserPrincipalId()
+
+      const updatedToken = await options[originalIndex].init(
+        Principal.fromText(publicKey),
+      )
+
+      setTokensOptions((prev) => {
+        const newOptions = prev.map((token, i) =>
```

Comment:
not critical, but map is bad usage here, its used to map array to change its values, in your case just copy array and change value by index